### PR TITLE
Fix item stacks qdeling themselves in Init and some item stack self-merge scenarios.

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

Stops `zero_amount()` from calling `qdel(src)` by default - Renames the proc `is_zero_amount(delete_if_zero = TRUE)`. All previous calls to `zero_amount()` will now specify if they want the stack to be qdel'd via the arg if the stack is zero. This is in place to stop a ton of duplicate is_cyborg checks while making sure there is no change in behaviour after the PR is merged.

Adds missing `check` arg to overridden `use()` procs to appease linter.

These changes were needed to support the fix for the below issue:

image
image

Item stacks no longer qdel(src) on Initialise.

Adds a new merge proc, `merge_without_del()` which merges two stacks and does not call a code path including `qdel(src)`. This can leave one of the stacks with an amount of zero. This is documented. This new merge proc is used in item stack `Initialize()` and if after this `is_zero_amount()` is TRUE, Initialize() returns INITIALIZE_HINT_QDEL to qdel itself gracefully.

As part of this, `can_merge()` no longer silently returns on error states where either stack is QDELETED or `target == src` and now CRASHes instead.

The old merge proc now calls this new merge_without_del proc and then calls `is_zero_amount()`. This preserves existing behaviour outside of the Initialise() edge case.

In testing, the change revealed that there was a signal handler that could potentially merge a stack into itself. This behaviour has been fixed by the below change:

Renames `/obj/item/stack/proc/on_entered(datum/source, atom/movable/crossing)` signal handler to `/obj/item/stack/proc/on_movable_entered_occupied_turf(datum/source, atom/movable/arrived)` which actually explains what event the signal handler is handling.

I also cover an edge case where the atom that enters the turf that src is on, is src itself. This is expected behaviour from the signal and needs bespoke handling.

Does a very minor bit of code cleanup for some procs I touched.

## Why It's Good For The Game

Better code goes brrrrr.

There are no player-facing changes. Old functionality should match new functionality, just with fewer stack_traces and CRASHes. Probably.

### CODE REVIEW

**Comment by Timberpoes** [comment]
https://discord.com/channels/326822144233439242/326831214667235328/875586481824100375

Todo sometime tomorrow.

File: code/game/objects/items/stacks/stack.dm [hunk 1]
```diff
@@ -391,11 +394,11 @@
 
 	return TRUE
 
-/obj/item/stack/proc/zero_amount()
+/// Returns TRUE if the item stack is the equivalent of a 0 amount item.
+/obj/item/stack/proc/is_zero_amount()
```
Comment (AnturK) [inline code/game/objects/items/stacks/stack.dm:398]: Won't this cause cyborg stacks to get deleted in all these ?

**Comment by AMonkeyThatCodes** [comment]
self-merging is against the rules, can you remove that? Else I might have to call my good friend @Iamgoofball into this PR.

**Comment by Iamgoofball** [comment]
selfmerging is against the rules

**Comment by Timberpoes** [comment]
Merging #60678 into this PR on https://github.com/Timberpoes/tgstation-nxt/tree/stack-feex-debug another branch seems to have stopped the unit test failures, so I'm waiting on the result of #60678 before this is good to go.

**Comment by LemonInTheDark** [comment]
@Timberpoes chop chop

File: code/game/objects/items/stacks/stack.dm (hunk 1 above)
Comment (Timberpoes) [reply code/game/objects/items/stacks/stack.dm:398]: Yes it will. I've come up with another solution which is a bit dirty, but gets the job done.

**Review by spookydonut — APPROVED** [review]

**Pull request merged**